include/graphblas/reference/blas3.hpp

```diff
@@ -1111,92 +1118,545 @@ namespace grb {
+					!is_object< InputType2 >::value &&
+					is_monoid< Monoid >::value
+				>::type * const = nullptr
+		) {
+
+#ifdef _DEBUG
+			std::cout << "In grb::internal::eWiseApply_matrix_generic_union\n";
+#endif
+			assert( phase != TRY );
+			constexpr bool crs_only = descr & descriptors::force_row_major;
+			// get whether the matrices should be transposed prior to execution
+			constexpr bool trans_left = descr & descriptors::transpose_left;
+			constexpr bool trans_right = descr & descriptors::transpose_right;
+
+			if( crs_only && (trans_left || trans_right) ) {
+#ifdef _DEBUG
+				std::cerr << "grb::descriptors::force_row_major and "
+					<< "grb::descriptors::transpose_left/right are mutually "
+					<< "exclusive\n";
+#endif
+				return ILLEGAL;
+			}
+
+			if( getID(A) == getID(C) || getID(B) == getID(C) ) {
+#ifdef _DEBUG
+				std::cerr << "grb::eWiseApply: The output matrix can not simultaneously be "
+					<< "one of the input matrices\n";
+#endif
+			}
+
+			// run-time checks
+			const size_t m = nrows( C );
+			const size_t n = ncols( C );
+			const size_t m_A = !trans_left ? nrows( A ) : ncols( A );
+			const size_t n_A = !trans_left ? ncols( A ) : nrows( A );
+			const size_t m_B = !trans_right ? nrows( B ) : ncols( B );
+			const size_t n_B = !trans_right ? ncols( B ) : nrows( B );
+
+			// Identities
+			const auto identity_A = monoid.template getIdentity< OutputType >();
+			const auto identity_B = monoid.template getIdentity< OutputType >();
+
+			if( m != m_A || m != m_B || n != n_A || n != n_B ) {
 #ifdef _DEBUG
-						std::cout << "A( " << i << ", " << k_col << " ) = " << A_raw.getValue( k,
-							mulMonoid.template getIdentity< typename Operator::D1 >() ) << ", ";
+				std::cerr << "grb::eWiseApply: dimensions of input matrices do not match\n";
 #endif
+				return MISMATCH;
+			}
+
+			const auto oper = monoid.getOperator();
+			const auto &A_raw = !trans_left ?
+				internal::getCRS( A ) :
+				internal::getCCS( A );
+			const auto &B_raw = !trans_right ?
+				internal::getCRS( B ) :
+				internal::getCCS( B );
+			auto &CRS_raw = internal::getCRS( C );
+			auto &CCS_raw = internal::getCCS( C );
+
+
+			// retrieve buffers
+			char *arr1 = nullptr, *arr3 = nullptr;
+			char *buf1 = nullptr, *buf3 = nullptr;
+			InputType1 * vbuf1 = nullptr;
+			OutputType * vbuf3 = nullptr;
+			internal::getMatrixBuffers( arr1, buf1, vbuf1, 1, A );
+			internal::getMatrixBuffers( arr3, buf3, vbuf3, 1, C );
+			// end buffer retrieval
+
+			// initialisations of the coordinates
+			// Note: By using the buffer of the output matrix C, we can
+			//       allow A and B to be the same matrix (with the same buffer)
+			Coordinates< reference > coors1, coors2;
+			coors1.set( arr1, false, buf1, n );
+			coors2.set( arr3, false, buf3, n );
+			// end initialisations of the coordinates
+
+			if( !crs_only ) {
+#ifdef _H_GRB_REFERENCE_OMP_BLAS3
+				#pragma omp parallel for simd default(none) shared(CCS_raw) firstprivate(n)
```

Comment:
make it (explicitly) statically scheduled (suggest to use localRange), no firstprivate on read-only n